integration/client/client_test.go

```diff
@@ -234,6 +234,93 @@ func TestImagePull(t *testing.T) {
 	}
 }
 
+// Image Pull should succeed when interrupted and restarted by upstream client
+// Ref: https://github.com/containerd/containerd/issues/3787
+func TestImagePullWithInterrupts(t *testing.T) {
+	client, err := newClient(t, address)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	testImage := imagelist.Get(imagelist.ResourceConsumer)
+
+	imageNameCh1 := make(chan string, 1)
+	imagePullErrCh1, getImageErrCh1 := make(chan error, 1), make(chan error, 1)
+	defer close(imageNameCh1)
+	defer close(imagePullErrCh1)
+	defer close(getImageErrCh1)
+
+	imageNameCh2 := make(chan string, 1)
+	imagePullErrCh2, getImageErrCh2 := make(chan error, 1), make(chan error, 1)
+	defer close(imageNameCh2)
+	defer close(imagePullErrCh2)
+	defer close(getImageErrCh2)
+
+	ctx1, cancel1 := testContext(t)
+	ctx2, cancel2 := testContext(t)
+
+	pullFn := func(ctx context.Context, shouldErr bool, imageNameCh chan<- string, imagePullErrCh chan<- error, getImageErrCh chan<- error) {
+		_, err = client.Pull(ctx, testImage, WithPullUnpack, WithPlatformMatcher(platforms.Default()))
+		imagePullErrCh <- err
+
+		img, err := client.GetImage(ctx, testImage)
+		getImageErrCh <- err
+		if shouldErr {
+			imageNameCh <- ""
```

Comment:
I think the `shouldErr` case is irrelevant here; ether the err happens (and you `imageNameCh <- ""` in the conditional below) or it _doesn't_ happen in which case there is an `img.Name()`.  We can assert whether an error is expected or not when we check the result.